ocaml/xapi/tar_ext.ml

```diff
@@ -0,0 +1,160 @@
+(*
+ * Copyright (c) Cloud Software Group, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Helpers
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+open D
+
+let dir_perm = 0o755
+
+let dir_size = 4096L
+
+let inode_size = 4096L
+
+let ( ++ ) = Int64.add
+
+let ( // ) = Filename.concat
+
+type unpack_error =
+  | Illegal_file_path of string
+  | Unsupported_file_type of string
+  | Unpacked_exceeds_max_size_limit of Int64.t
+  | File_size_mismatch of {
+        path: string
+      ; expected_size: Int64.t
+      ; actual_size: Int64.t
+    }
+  | File_incomplete
+  | File_corrupted
+  | Unpacking_failure
+
+let unpack_error_to_string = function
+  | Illegal_file_path path ->
+      Printf.sprintf "Illegal file path: %s" path
+  | Unsupported_file_type t ->
+      Printf.sprintf "Unsupported file type: %s" t
+  | Unpacked_exceeds_max_size_limit size ->
+      Printf.sprintf
+        "Stop unpacking, otherwise unpacked files exceed max size limit: %s"
+        (Int64.to_string size)
+  | File_size_mismatch {path; expected_size; actual_size} ->
+      Printf.sprintf
+        "Unpacked file size mismatch, path: %s, expected size: %s, actual \
+         size: %s"
+        path
+        (Int64.to_string expected_size)
+        (Int64.to_string actual_size)
+  | File_incomplete ->
+      "File incompete"
+  | File_corrupted ->
+      "File corrupted"
+  | Unpacking_failure ->
+      "Unpacking failure"
+
+type unpack_result = Next of Int64.t
+
+let unpack dir max_size_limit acc_size ifd head =
+  let assert_file_path_is_legal file_path =
+    if
+      List.exists
+        (fun s ->
+          String.equal Filename.current_dir_name s
+          || String.equal Filename.parent_dir_name s
+        )
+        (String.split_on_char '/' file_path)
+    then
+      Error (Illegal_file_path file_path)
+    else
+      Ok ()
+  in
+  let ( let* ) = Result.bind in
+  let assert_file_not_exceed_max_size acc_size =
+    if Int64.compare acc_size max_size_limit > 0 then
+      Error (Unpacked_exceeds_max_size_limit max_size_limit)
+    else
+      Ok ()
+  in
+  let file_name = head.Tar.Header.file_name in
+  debug "%s: Unpacking tar file: %s" __FUNCTION__ file_name ;
+  let* () = assert_file_path_is_legal file_name in
+  let file_path = dir // file_name in
+  match head.Tar.Header.link_indicator with
+  | Tar.Header.Link.Normal ->
+      let file_size = head.Tar.Header.file_size in
```

Comment:
In the case of using the record, the name can be `expected_size` directly. And then the `copied_size` can be `actual_size`.